### PR TITLE
add environ()

### PR DESCRIPTION
**Pull request by mattn**
This pull-request add getenv() to add a way to get environment variable or variables. `getenv("FOO")` works as same as  $FOO. `getenv()` return dictionary of pairs of key and value. If you have interest in this, I'll add test. But for the test, https://github.com/vim/vim/pull/2855 is required.

### CODE REVIEW

**Comment by mattn** [comment]
Do you think that this should return v:null for `getenv("NOT_EXISTS")`?

**Comment by mattn** [comment]
Changed to return v:null for non-existence variable. And added test. But now will fail. This will be succeeded  if #2855 will be merged.

**Comment by brammool** [comment]
Yasuhiro Matsumoto wrote:

> This pull-request add getenv() to add a way to get environment
> variable or variables. `getenv("FOO")` works as same as  $FOO.
> `getenv()` return dictionary of pairs of key and value. If you have
> interest in this, I'll add test. But for the test,
> https://github.com/vim/vim/pull/2855 is required.

Since $VAR already works to get the value of an environment variable,
does it make sense to have another way with getenv('VAR')?
Perhaps getenv() should just always return a dictionary with all of
them.


-- 
WOMAN:   I didn't know we had a king. I thought we were an autonomous
         collective.
DENNIS:  You're fooling yourself.  We're living in a dictatorship.  A
         self-perpetuating autocracy in which the working classes--
WOMAN:   Oh there you go, bringing class into it again.
DENNIS:  That's what it's all about if only people would--
                                  The Quest for the Holy Grail (Monty Python)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mattn** [comment]
@brammool please look my last change. We will be possible to check the existence of environment variable.

**Comment by brammool** [comment]
> @brammool please look my last change. We will be possible to check the
> existence of environment variable.

Well, but on most systems an empty environment variable should be
handled the same as non-existence.  I don't see much point in adding
something where the difference matters.  Seems like a theoretical
difference without practical use.

If getenv() returns a dict with all environment variables, then if you
really want you can tell the difference.  It's more code, but then it
would almost never be needed.

-- 
Individualists unite!

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mattn** [comment]
@brammool Okay, So, do you want to rename this to `environ()` ?

**Comment by brammool** [comment]
> @brammool Okay, So, do you want to rename this to `environ()` ?

I suppose that's better than getenv(), which has a meaning in most
programs.  In Unix "environ" is a variable, but using that for a
function name should be OK.  In Python it's os.environ.

Alternative is environment().  Hmm, perhaps that's better?  Or we could
keep it short: env().

-- 
ARTHUR:  Then who is your lord?
WOMAN:   We don't have a lord.
ARTHUR:  What?
DENNIS:  I told you.  We're an anarcho-syndicalist commune.  We take it in
         turns to act as a sort of executive officer for the week.
                                  The Quest for the Holy Grail (Monty Python)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mattn** [comment]
Now renamed to environ()

**Comment by k-takata** [comment]
This might be used for a workaround for the following todo item:

> Can't recognize the $ProgramFiles(x86) environment variable.  Recognize it
> specifically?  First try with the parens, then without.

**Comment by brammool** [comment]
Ken Takata wrote:

> This might be used for a workaround for the following todo item:
> 
> > Can't recognize the $ProgramFiles(x86) environment variable.  Recognize it
> > specifically?  First try with the parens, then without.

You mean to pass the name to the getenv() function?
It's a very specific usage, but I suppose it's better than trying to
parse the name when used in $VAR form.

We could still use environ() to get the dictionary and getenv(name) to
get the value of one.  Main reason is that the type of the return value
differs.

-- 
    f y cn rd ths thn y cn hv grt jb n cmptr prgrmmng

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by k-takata** [comment]
Yeah, using `getenv('ProgramFiles(x86)')` might be easier than using `environ()['ProgramFiles(x86)']` to get the value of the `ProgramFiles(x86)` environment variable.
(If `getenv()` is added, someone might also want `setenv()` for symmetricity. I'm not sure it is really needed.)

**Comment by mattn** [comment]
I changed `environ()` to use win32 wide string APIs since reflected that #4008 will be merged.

File: src/evalfunc.c [hunk 1]
```diff
@@ -5160,6 +5164,26 @@ f_getcwd(typval_T *argvars, typval_T *rettv)
...
+{
+    int	    mustfree = FALSE;
+    char_u  *p = vim_getenv(tv_get_string(&argvars[0]), &mustfree);
+    if (p == NULL)
+    {
+	rettv->v_type = VAR_SPECIAL;
+	rettv->vval.v_number = VVAL_NULL;
+	return;
+    }
+    if (mustfree)
```
Comment (k-takata) [inline src/evalfunc.c]: Shouldn't be `!mustfree`?

File: runtime/doc/eval.txt [hunk 2]
```diff
@@ -4849,6 +4851,11 @@ getcwd([{winnr} [, {tabnr}]])
 		the window in the specified tab page.
 		Return an empty string if the arguments are invalid.
 
+getenv([{name}])					*getenv()*
+		Return environment variable of {name}. See also |expr-env|.
+		If {name} is not given, return all of environment variables
+		as dictionary.
```
Comment (k-takata) [inline runtime/doc/eval.txt]: Are you going to implement this later?
It seems that this is not implemented yet.

Comment (k-takata) [reply runtime/doc/eval.txt]: Sorry, this was outdated.

File: src/evalfunc.c (hunk 1 above)
Comment (mattn) [reply src/evalfunc.c]: Right!

File: src/evalfunc.c [hunk 3]
```diff
@@ -3197,6 +3203,65 @@ f_empty(typval_T *argvars, typval_T *rettv)
...
+    do
+    {
+# ifdef MSWIN
+	if ((p = (short_u*)_wenviron[i]) == NULL)
+	    return;
+	namevalue = utf16_to_enc((short_u *)p, NULL);
+	if (namevalue == NULL)
+	    return;
+	if ((value = vim_strchr(namevalue, '=')) == NULL)
+	    return;
```
Comment (k-takata) [inline src/evalfunc.c]: `namevalue` leaks here.

File: src/evalfunc.c [hunk 4]
```diff
@@ -3197,6 +3203,65 @@ f_empty(typval_T *argvars, typval_T *rettv)
...
+	name = vim_strnsave(namevalue, value - namevalue);
+	vim_free(namevalue);
```
Comment (k-takata) [inline src/evalfunc.c]: If you free `namevalue` here, you cannot use `value` at the line 3258.

File: src/evalfunc.c [hunk 5]
```diff
@@ -3197,6 +3203,65 @@ f_empty(typval_T *argvars, typval_T *rettv)
...
+	    return;
+	if ((name = vim_strnsave(namevalue, value - namevalue)) == NULL)
+	    return;
+# endif
+	if ((value = vim_strsave(value + 1)) == NULL)
+	{
+	    vim_free(name);
+	    return;
+	}
+	dict_add_string(rettv->vval.v_dict, (char*)name, value);
```
Comment (k-takata) [inline src/evalfunc.c:3250]: I think `name` should be freed after here.

File: src/testdir/test_environ.vim [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+function Test_external_env()
+  call setenv('FOO', 'こんにちわ')
+  if has('win32')
+    let result = system('echo %FOO%')
+  else
+    let result = system('echo $FOO')
+  endif
+  let result = substitute(result, '[ \r\n]', '', 'g')
+  let result = iconv(result, 'char', &encoding)
+  call assert_equal(result, 'こんにちわ')
```
Comment (k-takata) [inline src/testdir/test_environ.vim]: Actual and expected are reversed.
BTW, I don't think that this test works on English version of Windows.

File: src/testdir/test_environ.vim [hunk 7]
```diff
@@ -25,15 +25,15 @@ function Test_setenv()
...
 function Test_external_env()
-  call setenv('FOO', 'こんにちわ')
+  call setenv('FOO', 'HelloWorld')
   if has('win32')
     let result = system('echo %FOO%')
   else
     let result = system('echo $FOO')
   endif
   let result = substitute(result, '[ \r\n]', '', 'g')
   let result = iconv(result, 'char', &encoding)
```
Comment (k-takata) [inline src/testdir/test_environ.vim]: Now `iconv` can be removed.

File: src/testdir/test_environ.vim [hunk 8]
```diff
@@ -0,0 +1,44 @@
+scriptencoding utf-8
+
+function Test_environ()
```
Comment (k-takata) [inline src/testdir/test_environ.vim]: Most other tests uses `func` instead of `function`.
(Same for other functions.)

**Pull request closed without merging**